runtime/service.go

```diff
@@ -318,6 +321,7 @@ func (s *service) StartShim(shimCtx context.Context, containerID, containerdBina
 	_, err = fcControlClient.CreateVM(shimCtx, &proto.CreateVMRequest{
 		VMID:                  s.vmID,
 		ExitAfterAllTasksGone: exitAfterAllTasksGone,
+		ContainerCount:        containerCount,
```

Comment:
That's done at `CreateVM`. `CreateVM` has a little odd bit of logic where it'll only create the VM once using sync. So basically we kind of abuse that here and call `CreateVM` but it may not create one. In the `StartShim` call we assume that either create VM has already been called or you are expecting only the call the shim and create a single container.

Reply:
I feel like we already have an indication of whether `CreateVM` should have been called or not: the presence or absence of the VM ID.  If we're only creating the VM once anyway (as we should), we probably should only call `CreateVM` if we're inside the conditional where VM ID is blank.

Reply:
If you *do* want to keep calling `CreateVM` here, you need a much more explicit comment in the code explaining why the value for `ContainerCount` is either explicitly `0` or `1` and the condition under which the value is ignored.

Reply:
I think we want to call `CreateVM` unconditionally. `StartShim` is called by containerd every single time a Task is going to be spawned. `StartShim`'s job is to determine if there is a shim for the incoming Task already and, if not, create a shim for the task.

In our firecracker-containerd world, creating a shim and creating a VM is the same operation. `StartShim` just always calls CreateVM when a new task is coming in to make sure there is a shim/VM for the task, interpreting an `AlreadyExists` error to mean there already is a shim/VM. We avoid doing something like "Check if VM already exists, then create it if not" in two separate operations because that'd open us to more TOCTTOU bugs to workaround.

+1 to adding more comments explaining the ContainerCount value setting though.

Reply:
@sipsma If we call CreateVM here with a ContainerCount of 0 and the VM has not already been created, do we end up with a VM that's unusable for running containers?

Reply:
I guess we do because of https://github.com/firecracker-microvm/firecracker-containerd/pull/154/files#diff-5728a8df56a2ca442bee8216792fbb6fR624?

Reply:
Yeah the default setting you linked to handles the case of ContainerCount being set to 0. The previous discussion around this is here: https://github.com/firecracker-microvm/firecracker-containerd/pull/154#discussion_r290004141